File: day15/Dia15.hs
module Main where

import Data.Bits
import Data.Word (Word64)

factor_a :: Word64
factor_a = 16807

factor_b :: Word64
factor_b = 48271

divisor :: Word64
divisor = 2147483647

initial_a :: Word64
initial_a = 516

initial_b :: Word64
initial_b = 190

last16bitsMask :: Word64
last16bitsMask = 0xffff

last16zeros :: Word64
last16zeros = 0x0000

produce :: Word64 -> Word64 -> Word64 -> Word64
produce factor divisor previous =
  rem (previous * factor) divisor

generator_a :: [Word64]
generator_a = go initial_a
  where
    go :: Word64 -> [Word64]
    go prev = let x = produce factor_a divisor prev
              in x : go x

generator_b :: [Word64]
generator_b = go initial_b
  where
    go :: Word64 -> [Word64]
    go prev = let x = produce factor_b divisor prev
              in x : go x

judge :: Word64 -> Word64 -> Bool
judge x y = (xor x y) .&. last16bitsMask == last16zeros

judgeNtimes :: Int -> Integer
judgeNtimes n = go n 0 generator_a generator_b
  where
    go :: Int -> Integer -> [Word64] -> [Word64] -> Integer
    go 0 acc _ _ = acc
    go n acc (a : as) (b : bs) =
      if judge a b
        then go (pred n) (acc + 1) as bs
        else go (pred n) acc as bs

main :: IO ()
main = do
  let res = judgeNtimes 40000000
  print res
  pure ()

File: day15/Dia15.rs
struct Generator {
    factor: i32,
    divisor: i32,
    previous: i32,
    mask: i32,
}

impl Generator {
    fn produce(&mut self) -> i32 {
        let result = (self.previous as i64 * self.factor as i64) % self.divisor as i64;
        self.previous = result as i32;
        self.previous
    }

    fn produce2(&mut self) -> i32 {
        let val = self.produce();
        if val & self.mask == 0 {
            val
        } else {
            self.produce2()
        }
    }
}

fn judge(a: i32, b: i32) -> bool {
    ((a ^ b) & 0xffff) == 0
}

fn main() {
    let mut gen_a = Generator{
        factor: 16807,
        divisor: 2147483647,
        // previous: 65,
        previous: 516,
        mask: (1 << 0) | (1 << 1),
    };
    let mut gen_b = Generator{
        factor: 48271,
        divisor: 2147483647,
        // previous: 8921,
        previous: 190,
        mask: (1 << 0) | (1 << 1) | (1 << 2),
    };

    let mut count = 0;
    let mut res_a;
    let mut res_b;

    // for _ in 0..40_000_000 {
    //     res_a = gen_a.produce();
    //     res_b = gen_b.produce();
    //     // println!("{:32b}", res_a);
    //     // println!("{:32b}", res_b);
    //     // println!("{}", (res_a ^ res_b) & 0xffff);
    //     // println!("{}", judge(res_a, res_b));
    //     // println!("");
    //     if judge(res_a, res_b) {
    //         count += 1;
    //     }
    // }
    for _ in 0..5_000_000 {
        res_a = gen_a.produce2();
        res_b = gen_b.produce2();
        if judge(res_a, res_b) {
            count += 1;
        }
    }
    println!("{}", count);
}
